### ISSUE TITLE
left-hand side indexing with a base case

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/cryptol#635]
```
fib = [0, 1] # fibs
fibs @ i = fib@i + fib@(i+1)
```

Here we get some peculiar results.

```
Main> fib : [2]_

[error] at <interactive>:1:1--1:11:
  Unsolvable constraint: 0 >= 1
Main> fib : [3]_
Showing a specific instance of polymorphic result:
  * Using 'Integer' for 2nd type argument of 'Main::fib'
[0, 1, 1]
Main> fib : [4]_
Showing a specific instance of polymorphic result:
  * Using 'Integer' for 2nd type argument of 'Main::fib'
[0, 1, 1, 1]
Main> fib : [5]_
Showing a specific instance of polymorphic result:
  * Using 'Integer' for 2nd type argument of 'Main::fib'
[0, 1, 1, 2, 3]
Main> fib : [9]_
Showing a specific instance of polymorphic result:
  * Using 'Integer' for 2nd type argument of 'Main::fib'
[0, 1, 1, 2, 3, 5, 8, 13, 21]
Main> fib : [10]_
Showing a specific instance of polymorphic result:
  * Using 'Integer' for 2nd type argument of 'Main::fib'
[0, 1, 1, 2, 3, 5, 8, 13, 21, 13]
```

I think I understand these oddities, having to do with `i` defaulting silently to be smaller than what I want.

This begs for left-hand side indexing to _properly_ support a base case. Any ideas?

**Comment by brianhuffman** [comment]
The first error message, which complains about "Unsolvable constraint: 0 >= 1", is due to the type that I gave to the `generate` primitive, which requires an array length of at least 1:
```
generate : {n, ix, a} (fin ix, n >= 1, ix >= width (n - 1)) => ([ix] -> a) -> [n]a
generate f = [ f i | i <- [0 .. n-1] ]
```
I could probably define `generate` in a different way that would allow `n = 0`. I didn't bother doing this before because the above definition of `generate` seems like the most natural one, and I wasn't sure whether allowing zero-length arrays would be worth the trouble of complicating it. Perhaps I should do it after all.

The weirdness with `fib : [10]_` is, as you suspected, due to the index type defaulting to `[3]` and the `i+1` overflowing. This is really not specific to having a "base case"; I ran into a very similar problem recently with a definition something like
```
xs @ i = ys@(2*i) # ys@(2*i+1)
```
where there is some computation on indices that is likely to overflow if the index type is defaulted to the smallest possible one for the array size.

Probably the best way to fix the problem would be to implement #502, so we could just use `Integer` for the index type. For now, you might want to get into the habit of always putting type constraints on your lhs indexes. Alternatively, you can rearrange your definition to use only subtraction (which can't overflow) instead of addition on the index, like this:
```
fibs @ i =
  if i < 2 then [0, 1] @ i else fibs@(i-2) + fibs@(i-1)
```

**Comment by brianhuffman** [comment]
Implementing #502 should take care of the problem of overflowing index calculations in this kind of definitions. I'll consider this ticket to be subsumed by #502.
